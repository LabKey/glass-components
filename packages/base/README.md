# @glass/base

Base components, models, and utility functions for LabKey applications and pages

## Release Notes ##

### version ???
*Released*: ???
* port FileAttachmentFormModel and Footer from Biologics
* add getDomainColumns() to AssayDefinitionModel
* export SubMenuItemProps and ISubItem for Biologics use cases

### version 0.9.0
*Released*: 17 July 2019
* Add LabelHelpTip as a general component for adding help tips and popovers next to a form label

### version 0.8.2
*Released*: 10 July 2019
* Modify ConfirmModal properties so most are optional

### version 0.8.1
*Released*: 10 July 2019
* Export misc assay models and date utility functions for Biologics usages (AssayDomainTypes, AssayLink, AssayUploadTabs, getDateFormat, isSampleLookup)
* Port recent changes to glass components from Biologics dev and bug fixes (SubMenuItem popover key, AssayDefinitionModel functions for getting sample columns, add WhereFilterType) 

### version 0.8.0
*Released*: 8 July 2019
* Add DesignSampleSetPermission to PermissionTypes

### version 0.7.0
*Released*: 3 July 2019
* Add unorderedEqual for checking of two arrays have the same elements in any order
* export dismissNotifications
* add optional parameters to SelectionMenuItem for maximum number of selected items and for specifying noun used in messages
* add method getCommonDataValues to create a map containing the data values shared across multiple rows of data (represented as a map)
* add getPkData method to QueryGridModel

### version 0.6.0
*Released*: 19 June 2019
* Include FilePreviewGrid functionality with FileAttachmentForm
* Add createGeneralAssayDesign, importGeneralAssayRun, and inferDomainFromFile actions and InferDomainResponse model
* port Cards (renamed from Dashboard) and WizardNavButtons from Biologics
* new ToggleButtons component
* add 'Download Template' button option to FileAttachmentForm component

### version 0.5.3
*Released*: 05 June 2019
* Use release version of @labkey/dependencies

### version 0.5.1
*Released*: 05 June 2019
* Update external package versions

### version 0.5.0
*Released*: 04 June 2019
* port SVGIcon and CreatedModified from Biologics
* port AssayDefinitionModel and fetchAllAssays from Biologics
* ManageDropdownButton and SelectionMenuItem

### version 0.4.0
*Released*: 03 June 2019
* port caseInsensitive, similaritySortFactory, spliceURL functions from Biologics
* port AssayProtocolModel and fetchProtocol model from Biologics
* export datePlaceholder function
* export capitalizeFirstChar utility method
* Add method inserting columns into QueryInfo

### version 0.3.0
*Released*: 03 June 2019
* Add ConfirmModal component

### version 0.2.0
*Released*: 22 May 2019
* PaginationButtons component and PaginationButtonsProps

### version 0.1.8
*Released*: 25 April 2019
* Add FileAttachmentForm and related components

### version 0.1.7
*Released*: 23 April 2019
* React bootstrap typings - augment @types/react-bootstrap and make shareable

### version 0.1.6
*Released*: 16 April 2019
* initial documented release


