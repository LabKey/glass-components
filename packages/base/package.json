{
  "name": "@glass/base",
  "version": "0.1.0",
  "description": "Base components, models, and utility functions for LabKey applications and pages",
  "main": "dist/base.cjs.js",
  "module": "dist/base.es.js",
  "types": "dist/index.d.ts",
  "files": [
    "dist/"
  ],
  "scripts": {
    "build": "nps build",
    "clean": "nps clean",
    "cleanAll": "nps cleanAll",
    "storybook": "start-storybook -p 9001 -c .storybook",
    "test": "cross-env NODE_ENV=test jest",
    "test_teamcity": "cross-env NODE_ENV=test jest --testResultsProcessor=jest-teamcity-reporter",
    "watch": "watch 'yarn build' ./src"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/LabKey/glass-components.git"
  },
  "author": "Cory Nathe",
  "license": "SEE LICENSE IN LICENSE.txt",
  "bugs": {
    "url": "https://github.com/LabKey/glass-components/issues"
  },
  "publishConfig": {
    "registry": "https://artifactory.labkey.com/artifactory/api/npm/libs-client"
  },
  "homepage": "https://github.com/LabKey/glass-components#readme",
  "dependencies": {
    "bootstrap": "3.3.7",
    "react-bootstrap": "0.32.4"
  },
  "peerDependencies": {
    "immutable": "^3.8.2",
    "react": "^16.0",
    "react-dom": "^16.0"
  },
  "jest": {
    "globals": {
      "LABKEY": {}
    },
    "moduleFileExtensions": [
      "ts",
      "tsx",
      "js"
    ],
    "moduleNameMapper": {
      "\\.(scss)$": "<rootDir>/../../test/styleMock.ts"
    },
    "setupFilesAfterEnv": [
      "<rootDir>/../../test/jest.setup.ts"
    ],
    "snapshotSerializers": [
      "enzyme-to-json/serializer"
    ],
    "testPathIgnorePatterns": [
      "/node_modules/"
    ],
    "testRegex": "(\\.(test|spec))\\.(ts|tsx)$",
    "transform": {
      "^.+\\.tsx?$": "ts-jest"
    }
  }
}
