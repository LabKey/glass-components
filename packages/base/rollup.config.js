/*
 * Copyright (c) 2018 LabKey Corporation
 *
 * Licensed under the Apache License, Version 2.0: http://www.apache.org/licenses/LICENSE-2.0
 */
import commonjs from 'rollup-plugin-commonjs';
import resolve from 'rollup-plugin-node-resolve';
import typescript from 'rollup-plugin-typescript2';
import sass from "rollup-plugin-sass";

const input = 'src/index.ts';

const globals = {
    'immutable': 'immutable',
    'react': 'React',
    'react-dom': 'ReactDOM'
};
const external = Object.keys(globals);

const namedExports = {
    // Named exports for packages
    // (when you get an error like: Error: 'getGlobal' is not exported by ../../node_modules/reactn/index.js)
    'reactn': ['getGlobal', 'setGlobal'],
};

export default [
    {
        external: external,
        input: input,
        output: {
            file: 'dist/base.cjs.js',
            format: 'cjs'
        },
        plugins: [
            resolve(),
            commonjs({namedExports}),
            typescript(),
            sass({
                output: 'dist/base.css'
            })
        ]
    },
    {
        external: external,
        input: input,
        output: {
            file: 'dist/base.es.js',
            format: 'es',
            name: 'base'
        },
        plugins: [
            resolve(),
            commonjs({namedExports}),
            typescript(),
            sass()
        ]
    },
    {
        external: external,
        input: input,
        output: {
            file: 'dist/base.umd.js',
            format: 'umd',
            name: 'base',
            globals: globals
        },
        plugins: [
            resolve(),
            commonjs({namedExports}),
            typescript(),
            sass()
        ]
    }
]
