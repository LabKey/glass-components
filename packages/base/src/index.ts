/*
 * Copyright (c) 2019 LabKey Corporation. All rights reserved. No portion of this work may be reproduced in
 * any form or by any electronic or mechanical means without written permission from LabKey Corporation.
 */
import { GRID_CHECKBOX_OPTIONS, GRID_EDIT_INDEX, GRID_SELECTION_INDEX, PermissionTypes } from './models/constants'
import {
    Container,
    IGridLoader,
    IGridResponse,
    IGridSelectionResponse,
    insertColumnFilter,
    IQueryGridModel,
    LastActionStatus,
    MessageLevel,
    QueryColumn,
    QueryGridModel,
    QueryInfo,
    QueryInfoStatus,
    QueryLookup,
    QuerySort,
    SchemaDetails,
    SchemaQuery,
    User,
    ViewInfo
} from './models/model'
import {
    applyDevTools,
    debounce,
    decodePart,
    devToolsActive,
    encodePart,
    generateId,
    getSchemaQuery,
    intersect,
    hasAllPermissions,
    naturalSort,
    not,
    resolveKey,
    resolveKeyFromJson,
    resolveSchemaQuery,
    toggleDevTools,
    toLowerSafe
} from './utils/utils'
import {
    AppURL,
    buildURL,
    getSortFromUrl,
    hasParameter,
    imageURL,
    setParameter,
    toggleParameter
} from './url/ActionURL'
import { Alert } from './components/Alert'
import { MultiMenuButton } from './components/menus/MultiMenuButton'
import { MenuOption, SubMenu } from "./components/menus/SubMenu";
import { SubMenuItem } from "./components/menus/SubMenuItem";
import { CustomToggle } from './components/CustomToggle'
import { LoadingSpinner } from './components/LoadingSpinner'
import { NotFound } from './components/NotFound'
import { Page, PageProps } from './components/Page'
import { LoadingPage, LoadingPageProps } from './components/LoadingPage'
import { PageHeader } from './components/PageHeader'
import { Progress } from './components/Progress'
import { Tip } from './components/Tip'
import { Grid, GridColumn, GridData, GridProps } from './components/Grid'
import { FormSection } from './components/FormSection'
import { FileAttachmentForm } from './components/FileAttachmentForm'
import { Notification } from './components/notifications/Notification'
import { createNotification } from './components/notifications/actions'
import { initNotificationsState } from './components/notifications/global'
import {
    MessageFunction,
    NotificationItemModel,
    NotificationItemProps,
    Persistence,
} from './components/notifications/model'
import {
    PermissionAllowed,
    PermissionNotAllowed,
} from "./components/Permissions"
import { PaginationButtons, PaginationButtonsProps } from './components/PaginationButtons';
// Import the scss file so it will be processed in the rollup scripts
import './theme/index.scss'

// export * from './typings/react-bootstrap.d.ts'

// Add explicit export block for the classes and functions to be exported from this package
export {
    // constants
    GRID_EDIT_INDEX,
    GRID_SELECTION_INDEX,
    GRID_CHECKBOX_OPTIONS,
    PermissionTypes,
    Persistence,

    // interfaces
    IQueryGridModel,
    IGridLoader,
    IGridResponse,
    IGridSelectionResponse,
    GridProps,
    LoadingPageProps,
    PageProps,

    //models
    AppURL,
    Container,
    User,
    QueryColumn,
    QueryGridModel,
    QueryInfo,
    QuerySort,
    QueryLookup,
    QueryInfoStatus,
    SchemaDetails,
    SchemaQuery,
    ViewInfo,
    MessageLevel,
    MessageFunction,
    NotificationItemProps,
    NotificationItemModel,
    LastActionStatus,
    GridColumn,
    GridData,

    //components
    Alert,
    CustomToggle,
    LoadingSpinner,
    LoadingPage,
    NotFound,
    Page,
    PageHeader,
    Progress,
    MenuOption,
    MultiMenuButton,
    Notification,
    SubMenu,
    SubMenuItem,
    Tip,
    Grid,
    PermissionAllowed,
    PermissionNotAllowed,
    PaginationButtons,
    PaginationButtonsProps,
    FormSection,
    FileAttachmentForm,

    // notification functions
    createNotification,
    initNotificationsState,

    // util functions
    decodePart,
    encodePart,
    getSchemaQuery,
    resolveKey,
    resolveKeyFromJson,
    resolveSchemaQuery,
    insertColumnFilter,
    intersect,
    hasAllPermissions,
    naturalSort,
    not,
    toLowerSafe,
    generateId,
    debounce,

    // url functions
    buildURL,
    getSortFromUrl,
    hasParameter,
    imageURL,
    setParameter,
    toggleParameter,

    // devTools functions
    applyDevTools,
    devToolsActive,
    toggleDevTools
}
