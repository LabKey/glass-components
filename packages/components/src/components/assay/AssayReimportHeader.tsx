import React from 'react';
import { Panel } from 'react-bootstrap';
import { Map } from 'immutable';
import { AppURL } from '../../url/AppURL';
import { AssayDefinitionModel } from '../base/models/model';

interface Props {
    hasBatchProperties?: boolean
    assay: AssayDefinitionModel
    replacedRunProperties: Map<string, any>
}

export class AssayReimportHeader extends React.Component<Props> {

    render() {
        const { assay, hasBatchProperties, replacedRunProperties } = this.props;
        let assayRunUrl = AppURL.create('assays', assay.type, assay.name, 'runs', replacedRunProperties.get('RowId'));
        return (
            <Panel>
                <Panel.Heading>
                    Re-Import Run
                </Panel.Heading>
                <Panel.Body>
                    <p>
                        <strong>Replacing Run: <a href={assayRunUrl.toHref()}>{replacedRunProperties.get('Name')}</a></strong>
                    </p>
                    <p>
                        Edit the {hasBatchProperties ? 'batch and ' : ''} run properties below or provide updated data for this assay run.  Changes will be reflected in the audit history for this run.
                    </p>
                </Panel.Body>
            </Panel>
        )
    }
}
