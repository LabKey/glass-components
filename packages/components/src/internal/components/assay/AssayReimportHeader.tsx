import React, { FC, memo } from 'react';

import { AssayDefinitionModel } from '../../AssayDefinitionModel';
import { caseInsensitive } from '../../util/utils';
import { AppURL } from '../../url/AppURL';

interface Props {
    assay: AssayDefinitionModel;
    hasBatchProperties?: boolean;
    replacedRunProperties: Record<string, any>;
}

export const AssayReimportHeader: FC<Props> = memo(({ assay, hasBatchProperties, replacedRunProperties }) => {
    const rowId = caseInsensitive(replacedRunProperties, 'RowId').value;
    const name = caseInsensitive(replacedRunProperties, 'Name').value;
    const assayRunUrl = AppURL.create('assays', assay.type, assay.name, 'runs', rowId);

    return (
        <div className="panel panel-default">
            <div className="panel-heading">Re-Import Run</div>
            <div className="panel-body">
                <p>
                    <strong>
                        Replacing Run: <a href={assayRunUrl.toHref()}>{name}</a>
                    </strong>
                </p>
                <p>
                    Edit the {hasBatchProperties ? 'batch and ' : ''} run properties below or provide updated data for
                    this assay run. Changes will be reflected in the audit history for this run.
                </p>
            </div>
        </div>
    );
});
