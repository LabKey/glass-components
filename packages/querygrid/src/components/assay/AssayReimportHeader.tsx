import * as React from 'react';
import { Panel } from 'react-bootstrap';
import { Map } from 'immutable';
import { AppURL, AssayDefinitionModel } from '@glass/base';

interface Props {
    hasBatchProperties?: boolean
    assay: AssayDefinitionModel
    replacedRunData: Map<string, any>
}

export class AssayReimportHeader extends React.Component<Props> {

    render() {
        const { assay, hasBatchProperties, replacedRunData } = this.props;
        let assayRunUrl = AppURL.create('assays', assay.type, assay.name, 'runs', replacedRunData.getIn(['RowId', 'value']));
        return (
            <Panel>
                <Panel.Heading>
                    Re-Import Run
                </Panel.Heading>
                <Panel.Body>
                    <p>
                        <strong>Re-importing Run: <a href={assayRunUrl.toHref()}>{replacedRunData.getIn(['Name', 'value'])}</a></strong>
                    </p>
                    <p>
                        Edit the {hasBatchProperties ? 'batch and ' : ''} run properties below or provide updated data for this assay run.  Changes will be reflected in the audit history for this run.
                    </p>
                </Panel.Body>
            </Panel>
        )
    }
}