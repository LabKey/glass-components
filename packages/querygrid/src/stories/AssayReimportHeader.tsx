/*
 * Copyright (c) 2019 LabKey Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import * as React from 'react';
import { fromJS } from "immutable";
import { storiesOf } from "@storybook/react";
import { boolean, withKnobs } from '@storybook/addon-knobs'
import { AssayDefinitionModel } from "@glass/base";
import './stories.scss'
import { AssayReimportHeader } from '../components/assay/AssayReimportHeader';

import assayDefJSON from '../test/data/assayDefinitionModel.json';

storiesOf('AssayReimportHeader', module)
    .addDecorator(withKnobs)
    .add("with knobs", () => {

        const assay = AssayDefinitionModel.create(assayDefJSON);

        const runData = fromJS({
            'RowId': "10",
            'Name':  'Test Name'
        });
        return (
            <AssayReimportHeader
                hasBatchProperties={boolean("Has batch properties?", false)}
                assay={assay}
                replacedRunProperties={runData}
            />
        )
    });