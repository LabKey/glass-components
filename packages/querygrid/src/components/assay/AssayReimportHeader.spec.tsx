import * as React from 'react';
import renderer from 'react-test-renderer'

import { AssayDefinitionModel } from '@glass/base';
import { fromJS } from 'immutable';
import { AssayReimportHeader } from './AssayReimportHeader';
import assayDefJSON from '../../test/data/assayDefinitionModel.json';


describe("<AssayReimportHeader/>", () => {
    const assay = AssayDefinitionModel.create(assayDefJSON);

    const runData = fromJS({
        'RowId': {
            'value': 10
        },
        'Name': {
            'value': 'Test Name'
        }
    });
    test("has batch properties", () => {
        const component = <AssayReimportHeader hasBatchProperties={true} assay={assay} replacedRunData={runData}/>
        const tree = renderer.create(component).toJSON();
        expect(tree).toMatchSnapshot();
    });

    test("no batch properties", () => {
        const component = <AssayReimportHeader hasBatchProperties={false} assay={assay} replacedRunData={runData}/>
        const tree = renderer.create(component).toJSON();
        expect(tree).toMatchSnapshot();
    })
});