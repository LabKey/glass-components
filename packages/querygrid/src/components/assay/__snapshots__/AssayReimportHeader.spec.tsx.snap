// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AssayReimportHeader/> has batch properties 1`] = `
<div
  className="panel panel-default"
>
  <div
    className="panel-heading"
  >
    Re-Import Run
  </div>
  <div
    className="panel-body"
  >
    <p>
      <strong>
        Re-importing Run: 
        <a
          href="#/assays/general/gpat%201/runs/10"
        >
          Test Name
        </a>
      </strong>
    </p>
    <p>
      Edit the 
      batch and 
       run properties below or provide updated data for this assay run.  Changes will be reflected in the audit history for this run.
    </p>
  </div>
</div>
`;

exports[`<AssayReimportHeader/> no batch properties 1`] = `
<div
  className="panel panel-default"
>
  <div
    className="panel-heading"
  >
    Re-Import Run
  </div>
  <div
    className="panel-body"
  >
    <p>
      <strong>
        Re-importing Run: 
        <a
          href="#/assays/general/gpat%201/runs/10"
        >
          Test Name
        </a>
      </strong>
    </p>
    <p>
      Edit the 
      
       run properties below or provide updated data for this assay run.  Changes will be reflected in the audit history for this run.
    </p>
  </div>
</div>
`;
